src/dataflow/src/sink/kafka.rs

```diff
@@ -574,3 +581,161 @@ where
+fn avro_encode_stream<G>(
+    input_stream: &Stream<G, ((Option<Row>, Option<Row>), u64, isize)>,
+    encoder: Encoder,
+    key_schema_id: Option<i32>,
+    value_schema_id: i32,
+    fuel: usize,
+    sink_name: String,
+) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), u64, isize)>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let name = format!("{}-avro_encode", sink_name);
+
+    let mut builder = OperatorBuilder::new(name, input_stream.scope());
+    let mut input = builder.new_input(&input_stream, Pipeline);
+    let (mut output, output_stream) = builder.new_output();
+    builder.set_notify(false);
+
+    let activator = input_stream
+        .scope()
+        .activator_for(&builder.operator_info().address[..]);
+
+    let mut stash: HashMap<Capability<Timestamp>, Vec<_>> = HashMap::new();
+    let mut vector = Vec::new();
+    let mut encode_logic = move |input: &mut FrontieredInputHandle<
+        Timestamp,
+        ((Option<Row>, Option<Row>), Timestamp, Diff),
+        _,
+    >,
+                                 output: &mut OutputHandle<
+        _,
+        ((Option<Vec<u8>>, Option<Vec<u8>>), Timestamp, Diff),
+        _,
+    >| {
+        let mut fuel_remaining = fuel;
+        // stash away all the input we get, we want to be a nice citizen
+        input.for_each(|cap, data| {
+            data.swap(&mut vector);
+            let stashed = stash.entry(cap.retain()).or_default();
+            for record in vector.drain(..) {
+                stashed.push(record);
+            }
+        });
+
+        // work off some of our data and then yield, can't be hogging
+        // the worker for minutes at a time
+
+        while fuel_remaining > 0 && !stash.is_empty() {
+            let lowest_ts = stash
+                .keys()
+                .min_by(|x, y| x.time().cmp(y.time()))
+                .expect("known to exist")
+                .clone();
+            let records = stash.get_mut(&lowest_ts).expect("known to exist");
+
+            let mut session = output.session(&lowest_ts);
+            let num_records_to_drain = cmp::min(records.len(), fuel_remaining);
+            // let num_records_to_drain = records.len();
+            records
+                .drain(..num_records_to_drain)
+                .for_each(|((key, value), time, diff)| {
+                    let key =
+                        key.map(|key| encoder.encode_key_unchecked(key_schema_id.unwrap(), key));
+                    let value =
+                        value.map(|value| encoder.encode_value_unchecked(value_schema_id, value));
+                    session.give(((key, value), time, diff));
+                });
+
+            fuel_remaining -= num_records_to_drain;
+
+            if records.is_empty() {
+                // drop our capability for this time
+                stash.remove(&lowest_ts);
+            }
+        }
+
+        if !stash.is_empty() {
+            activator.activate();
+            return true;
```

Comment:
nit: I don't think you need the return statement for correctness

Reply:
Yes, I left it in more for clarity, but I can remove. 👍

Reply:
ok so in the other case where we use `build_reschedule` in `sink_logic` in this file, whenever we return true to `build_reschedule` we also manually specify when we want to be activated
I'm looking at the following:
```
552         if !ready_rows.is_empty() {
553             // We need timely to reschedule this operator as we have pending
554             // items that we need to send to Kafka
555             s.activator.activate();
556             return true;
557         }
558
559         if in_flight > 0 {
560             // We still have messages that need to be flushed out to Kafka
561             // Let's make sure to keep the sink operator around until
562             // we flush them out
563             s.activator.activate_after(Duration::from_secs(5));
564             return true;
565         }
566
```

my understanding is that a return value of `true` indicates to timely "you do not have permission to delete this operator", but it doesn't actually reschedule the operator and the operator needs to use the activator to do so manually

Reply:
oh no hang on -- you can just do this

```
@@ -664,8 +663,8 @@ where
             activator.activate();
             return true;
         }
-        // signal that we're still incomplete
-        !stash.is_empty()
+        // signal that we're complete.
+        false
     };
     ```
  which i think is much simpler

Reply:
thanks! 👌